sim/midas/src/main/cc/simif_token_hashers.cc

```diff
@@ -0,0 +1,222 @@
+#include "simif_token_hashers.h"
+#include "simif.h"
+
+#include <iostream>
+#include <sstream>
+#include <fstream>
+
+/**
+ * The constructor for simif_token_hashers_t.
+ *
+ * @param [in] p A pointer to the parent instance of simif_t
+ */
+simif_token_hashers_t::simif_token_hashers_t(simif_t *p) : parent(p) {
+
+#ifdef TOKENHASHMASTER_0_PRESENT
+  TOKENHASHMASTER_0_substruct_create;
+  trigger0 = TOKENHASHMASTER_0_substruct->triggerDelay0_TokenHashMaster;
+  trigger1 = TOKENHASHMASTER_0_substruct->triggerDelay1_TokenHashMaster;
+  period0 = TOKENHASHMASTER_0_substruct->triggerPeriod0_TokenHashMaster;
+  period1 = TOKENHASHMASTER_0_substruct->triggerPeriod1_TokenHashMaster;
+  free(TOKENHASHMASTER_0_substruct);
+
+  cnt = TOKENHASH_COUNT;
+  bridge_names.assign(TOKENHASH_BRIDGENAMES, TOKENHASH_BRIDGENAMES + cnt);
+  names.assign(TOKENHASH_NAMES, TOKENHASH_NAMES + cnt);
+  outputs.assign(TOKENHASH_OUTPUTS, TOKENHASH_OUTPUTS + cnt);
+  queue_heads.assign(TOKENHASH_QUEUEHEADS, TOKENHASH_QUEUEHEADS + cnt);
+  queue_occupancies.assign(TOKENHASH_QUEUEOCCUPANCIES,
+                           TOKENHASH_QUEUEOCCUPANCIES + cnt);
+  tokencounts0.assign(TOKENHASH_TOKENCOUNTS0, TOKENHASH_TOKENCOUNTS0 + cnt);
+  tokencounts1.assign(TOKENHASH_TOKENCOUNTS1, TOKENHASH_TOKENCOUNTS1 + cnt);
+#endif
+}
+
+/**
+ * Print debug info about the MMIO internals
+ */
+void simif_token_hashers_t::info() {
+  std::cout << "trigger0 " << trigger0 << "\n";
+  std::cout << "trigger1 " << trigger1 << "\n";
+  std::cout << "period0 " << period0 << "\n";
+  std::cout << "period1 " << period1 << "\n";
+
+  for (uint32_t i = 0; i < cnt; i++) {
+    std::cout << "i: " << i << "\n";
+    std::cout << "  bridge: " << bridge_names[i] << "\n";
+    std::cout << "  name: " << names[i] << "\n";
+    std::cout << "  direction: " << (outputs[i] ? "Output" : "Input") << "\n";
+    std::cout << "  queue_head: " << queue_heads[i] << "\n";
+    std::cout << "  queue_occupancy: " << queue_occupancies[i] << "\n";
+    std::cout << "  tokencount0: " << tokencounts0[i] << "\n";
+    std::cout << "  tokencount1: " << tokencounts1[i] << "\n";
+  }
+}
+
+/**
+ * Set the delay and period for the token hashers
+ *
+ * @param [in] delay The number of tokens before hashes are saved
+ * @param [in] period The number of hashes to skip between saving. 0 means save
+ * every hash
+ */
+void simif_token_hashers_t::set_params(const uint64_t delay,
+                                       const uint64_t period) {
+#ifndef TOKENHASHMASTER_0_PRESENT
```

Comment:
All these macros should be hoisted out - if a token hasher is not present, we should not instantiate this class. We should not change the behaviour of the class based on whether a component is present or not.